[D1] Add new example to use D1 with Prisma ORM

nikolasburk opened the pull request:
A PR that:
- mentions Prisma ORM in community projects: https://developers.cloudflare.com/d1/reference/community-projects/
- adds step-by-step instructions to deploy a CF Worker that queries D1 using Prisma ORM as another example: https://developers.cloudflare.com/d1/examples/

Let mw know if you have any feedback and/or change requests, happy to make any adjustments!

elithrar commented:
Thanks @nikolasburk! We’ll review this week and get this published.

vy-ton commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+    "name" TEXT
+);
+
+-- CreateIndex
```
helpful to explain why this index was autogenerated.

nikolasburk replied:
added this note:

Note that the `UNIQUE INDEX` on the `email` was created because the `User` model in your Prisma schema is using the [`@unique`](https://www.prisma.io/docs/orm/reference/prisma-schema-reference#unique) attribute on its `email` field.

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+
+### 5. Query your database from the Worker
+
+In order to query your database from the Worker using Prisma ORM, you need to:
```
```suggestion
To query your database from the Worker using Prisma ORM, you need to:
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,281 @@
+
+Before running the Worker, you need to generate Prisma Client with the following command:
+
+```
````
```suggestion
```sh
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,281 @@
+Before running the Worker, you need to generate Prisma Client with the following command:
+
+```
+npx prisma generate
````
```suggestion
$ npx prisma generate
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+
+### 6. Run the Worker locally
+
+With the database query in place and Prisma Client generated, you can go ahead and run the Worker locally:
```
```suggestion
Now that you have the database query in place and Prisma Client generated, run the Worker locally:
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,281 @@
+
+With the database query in place and Prisma Client generated, you can go ahead and run the Worker locally:
+
+```
````
```suggestion
```sh
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,281 @@
+With the database query in place and Prisma Client generated, you can go ahead and run the Worker locally:
+
+```
+npm run dev
````
```suggestion
$ npm run dev
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,281 @@
+npm run dev
+```
+
+Now you can open your browser at [`http://localhost:8787`](http://localhost:8787/) to see the result of the database query:
````
```suggestion
Open your browser at [`http://localhost:8787`](http://localhost:8787/) to check the result of the database query:
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,281 @@
+
+To deploy the Worker, run the the following command:
+
+```
````
```suggestion
```sh
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,281 @@
+To deploy the Worker, run the the following command:
+
+```
+npm run deploy
````
```suggestion
$ npm run deploy
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,281 @@
+npm run deploy
+```
+
+As before, your deployed Worker is accessible via `https://prisma-d1-example.USERNAME.workers.dev`. If you navigate your browser to that URL, you should see the following data that's queried from your remote D1 database:
````
```suggestion
Access your Worker at `https://prisma-d1-example.USERNAME.workers.dev`. Your browser should display the following data queried from your remote D1 database:
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+
+Congratulations, you just deployed a Cloudflare Worker using D1 as a database and querying it via Prisma ORM 🎉
+
+## Next steps
```
```suggestion
## Related resources
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+
+## Next steps
+
+Here are some more useful resources to check out if you want to learn more about Prisma ORM:
```
```suggestion
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+
+Here are some more useful resources to check out if you want to learn more about Prisma ORM:
+
+- **Docs**: [Get started with Prisma ORM](https://www.prisma.io/docs/getting-started)
```
```suggestion
- [Prisma documentation](https://www.prisma.io/docs/getting-started).
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+Here are some more useful resources to check out if you want to learn more about Prisma ORM:
+
+- **Docs**: [Get started with Prisma ORM](https://www.prisma.io/docs/getting-started)
+- **Help**: Open a new [GitHub Discussion](https://github.com/prisma/prisma/discussions/) or [ask the AI bot in the Prisma docs](https://www.prisma.io/docs)
```
```suggestion
- To get help, open a new [GitHub Discussion](https://github.com/prisma/prisma/discussions/), or [ask the AI bot in the Prisma docs](https://www.prisma.io/docs).
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+
+- **Docs**: [Get started with Prisma ORM](https://www.prisma.io/docs/getting-started)
+- **Help**: Open a new [GitHub Discussion](https://github.com/prisma/prisma/discussions/) or [ask the AI bot in the Prisma docs](https://www.prisma.io/docs)
+- **Examples**: [Ready-to-run examples using Prisma ORM](https://github.com/prisma/prisma-examples/)
```
```suggestion
- [Ready-to-run examples using Prisma ORM](https://github.com/prisma/prisma-examples/).
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+- **Docs**: [Get started with Prisma ORM](https://www.prisma.io/docs/getting-started)
+- **Help**: Open a new [GitHub Discussion](https://github.com/prisma/prisma/discussions/) or [ask the AI bot in the Prisma docs](https://www.prisma.io/docs)
+- **Examples**: [Ready-to-run examples using Prisma ORM](https://github.com/prisma/prisma-examples/)
+- **Community**: Check out the [Prisma community](https://www.prisma.io/community), follow [Prisma on X](https://www.x.com/prisma) and join the [Prisma Discord](https://pris.ly/discord)
```
```suggestion
- Check out the [Prisma community](https://www.prisma.io/community), follow [Prisma on X](https://www.x.com/prisma) and join the [Prisma Discord](https://pris.ly/discord).
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+- **Help**: Open a new [GitHub Discussion](https://github.com/prisma/prisma/discussions/) or [ask the AI bot in the Prisma docs](https://www.prisma.io/docs)
+- **Examples**: [Ready-to-run examples using Prisma ORM](https://github.com/prisma/prisma-examples/)
+- **Community**: Check out the [Prisma community](https://www.prisma.io/community), follow [Prisma on X](https://www.x.com/prisma) and join the [Prisma Discord](https://pris.ly/discord)
+- **Blog**: [Developer Experience Redefined: Prisma & Cloudflare Lead the Way to Data DX](https://www.prisma.io/blog/cloudflare-partnership-qerefgvwirjq)
```
```suggestion
- [Developer Experience Redefined: Prisma & Cloudflare Lead the Way to Data DX](https://www.prisma.io/blog/cloudflare-partnership-qerefgvwirjq)
```

Maddy-Cloudflare commented on content/d1/reference/community-projects.md:
```diff
@@ -23,6 +23,15 @@ Two community tools are available for creating backups/exports of D1 databases a
 
+### Prisma ORM
+
+[Prisma ORM](https://www.prisma.io/orm) is a next-generation JavaScript and TypeScript ORM that unlocks a new level of developer experience when working with databases thanks to its intuitive data model, automated migrations, type-safety & auto-completion.
```
```suggestion
[Prisma ORM](https://www.prisma.io/orm) is a next-generation JavaScript and TypeScript ORM that unlocks a new level of developer experience when working with databases thanks to its intuitive data model, automated migrations, type-safety and auto-completion.
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,281 @@
+npx prisma migrate diff --from-empty --to-schema-datamodel ./prisma/schema.prisma --script > prisma/schema.sql
+```
+
+This stores a SQL statement to create a new `User` table in a new file called `schema.sql` in the `prisma` directory. Here's what it looks like:
````
```suggestion
This stores a SQL statement to create a new `User` table in a new file called `schema.sql` in the `prisma` directory. Your `schema.sql` should look like this:
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+
+### 7. Deploy the Worker
+
+To deploy the Worker, run the the following command:
```
```suggestion
To deploy the Worker, run the following command:
```

nikolasburk commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,281 @@
+
+[[d1_databases]]
+binding = "DB" # i.e. available in your Worker on env.DB
+database_name = "prisma-demo-db"
```
Ah sorry I missed this 👍

nikolasburk commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,281 @@
+CREATE UNIQUE INDEX "User_email_key" ON "User"("email");
+```
+
+You now need to use the `wrangler d1 execute`  command to send this SQL statement to D1. Note that this command accepts two options:
````
Ha, good callout! I adjusted the content to have the following flow:

1. create empty migration with `npx wrangler d1 migrations create prisma-demo-db create_user_table`
1. add `model User` to Prisma schema
1. generate SQL statement into migration file: `npx prisma migrate diff --from-empty --to-schema-datamodel ./prisma/schema.prisma --script > migrations/0001_create_user_table.sql`
1. apply migrations locally and remotely with `npx wrangler d1 migrations apply prisma-demo-db`

nikolasburk commented:
Wow, what an in-depth review! Thanks so much @Maddy-Cloudflare and @vy-ton 🙏 I addressed all of your feedback, let me know if there's anything else that should be changed 😄

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,308 @@
+CREATE UNIQUE INDEX "User_email_key" ON "User"("email");
+```
+
+Note that the `UNIQUE INDEX` on the `email` was created because the `User` model in your Prisma schema is using the [`@unique`](https://www.prisma.io/docs/orm/reference/prisma-schema-reference#unique) attribute on its `email` field.
````
```suggestion
`UNIQUE INDEX` on `email` was created because the `User` model in your Prisma schema is using the [`@unique`](https://www.prisma.io/docs/orm/reference/prisma-schema-reference#unique) attribute on its `email` field.
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,308 @@
+$ npx wrangler d1 migrations apply prisma-demo-db --remote
+```
+
+Hit **Enter** both times when you're prompted to confirm that the migration should be applied.
````
```suggestion
Select **Enter** both times when you are prompted to confirm that the migration should be applied.
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,308 @@
+
+```
+
+Before running the Worker, you need to generate Prisma Client with the following command:
````
```suggestion
Before running the Worker, generate Prisma Client with the following command:
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,308 @@
+[{"id":1,"email":"jane@prisma.io","name":"Jane Doe (Remote)"}]
+```
+
+Congratulations, you just deployed a Cloudflare Worker using D1 as a database and querying it via Prisma ORM 🎉
````
```suggestion
By finishing this tutorial, you have deployed a Cloudflare Worker using D1 as a database and querying it via Prisma ORM.
```

Maddy-Cloudflare commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,308 @@
+## Related resources
+
+- [Prisma documentation](https://www.prisma.io/docs/getting-started).
+-  To get help, open a new [GitHub Discussion](https://github.com/prisma/prisma/discussions/), or [ask the AI bot in the Prisma docs](https://www.prisma.io/docs).
```
```suggestion
- To get help, open a new [GitHub Discussion](https://github.com/prisma/prisma/discussions/), or [ask the AI bot in the Prisma docs](https://www.prisma.io/docs).
```

Maddy-Cloudflare commented on content/d1/reference/community-projects.md:
```diff
@@ -23,6 +23,15 @@ Two community tools are available for creating backups/exports of D1 databases a
+
+* [Tutorial](/d1/examples/d1-and-prisma-orm/)
+* [Docs](https://www.prisma.io/docs/orm/prisma-client/deployment/edge/deploy-to-cloudflare#d1)
+* Example (Coming soon)
```
```suggestion
* Tutorial (Coming soon)
```

nikolasburk replied:
I was referring to the deployment example project that we can link when next Tuesday when it gets merged: https://github.com/prisma/prisma-examples/pull/5793

nikolasburk replied:
but maybe we should drop it altogether while it's not there...

```suggestion
```

Maddy-Cloudflare replied:
> but maybe we should drop it altogether while it's not there...
Sorry then, my mistake. We can drop it if it's not there.

Maddy-Cloudflare commented on content/d1/reference/community-projects.md:
```diff
@@ -23,6 +23,15 @@ Two community tools are available for creating backups/exports of D1 databases a
+* [Tutorial](/d1/examples/d1-and-prisma-orm/)
+* [Docs](https://www.prisma.io/docs/orm/prisma-client/deployment/edge/deploy-to-cloudflare#d1)
+* Example (Coming soon)
+
```
```suggestion
```

Maddy-Cloudflare commented:
> Wow, what an in-depth review! Thanks so much @Maddy-Cloudflare and @vy-ton 🙏 I addressed all of your feedback, let me know if there's anything else that should be changed 😄

Thanks for creating this tutorial! I tested this and it works fine.

nikolasburk commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,308 @@
+title: Query D1 using Prisma ORM
+weight: 3
+layout: example
+---
```
Ah I see! I'll move it under `tutorials` then. FWIW, we'll be adding an actual deployment example with D1 to our `prisma-examples` repo as well. Maybe we can describe that in the `examples` directory.

nikolasburk commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,308 @@
+
+When prompted if the command can create a new folder called `migrations`, hit **Enter** to confirm.
+
+The command has now created a new directory called `migrations` and an empty file called `0001_create_user_table.sql` inside of it:
```
can you double-check this suggestion? I think the sentence becomes factually incorrect with it?

nikolasburk commented on content/d1/examples/d1-and-prisma-orm.md:
````diff
@@ -0,0 +1,308 @@
+
+The command has now created a new directory called `migrations` and an empty file called `0001_create_user_table.sql` inside of it:
+
+```
````
can you double-check this suggestion? the code snippet displays the structure on the file system and is not a sql code snippet.

nikolasburk commented on content/d1/examples/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,308 @@
++  id    Int     @id @default(autoincrement())
++  email String  @unique
++  name  String?
++}
```
they were intentional. in combination with the `diff` language annotation, it renders the lines prefixes with a `+` in green, e.g. this markdown:

````
```diff
+model User {
+  id    Int     @id @default(autoincrement())
+  email String  @unique
+  name  String?
+}
```
````

is rendered as followed:

```diff
+model User {
+  id    Int     @id @default(autoincrement())
+  email String  @unique
+  name  String?
+}
```

it's a small detail though and I think we can also remove them (since they may cause some friction when people directly copy the content of this code block).

nikolasburk commented on content/d1/reference/community-projects.md:
```diff
@@ -23,6 +23,14 @@ Two community tools are available for creating backups/exports of D1 databases a
+
+[Prisma ORM](https://www.prisma.io/orm) is a next-generation JavaScript and TypeScript ORM that unlocks a new level of developer experience when working with databases thanks to its intuitive data model, automated migrations, type-safety and auto-completion.
+
+* [Tutorial](/d1/examples/d1-and-prisma-orm/)
```
but the link to the tutorial needs to be updated to point to the `tutorials` directory I guess?

```suggestion
* [Tutorial](/d1/tutorials/d1-and-prisma-orm/)
```

nikolasburk commented:
Thanks again for the follow-up @Maddy-Cloudflare 🙏 I went ahead and incorporated all your suggestions (except for two of them, where I left a comment and asked you double-check because I think they would make the content incorrect).

Maddy-Cloudflare commented:
> Thanks again for the follow-up @Maddy-Cloudflare 🙏 I went ahead and incorporated all your suggestions (except for two of them, where I left a comment and asked you double-check because I think they would make the content incorrect).

It's good to be merged, thanks for applying the suggestions!

vy-ton commented:
@Maddy-Cloudflare Does the directory structure have to be `content/d1/tutorials/<folder>/index.md`?

vy-ton commented on content/d1/tutorials/d1-and-prisma-orm.md:
```diff
@@ -0,0 +1,305 @@
+
+Once you deploy your Worker, you should be able to preview your Worker at `https://prisma-d1-example.USERNAME.workers.dev`, which displays "Hello World" in the browser.
+
+### 2. Initialize Prisma ORM
```
@nikolasburk Can you mention the required minimum Prisma ORM version

nikolasburk replied:
Sure, here's a note:

```suggestion
### 2. Initialize Prisma ORM

> D1 is supported in Prisma ORM as of [v5.12.0](https://github.com/prisma/prisma/releases/tag/5.12.0).
```

Maddy-Cloudflare commented:
> @Maddy-Cloudflare Does the directory structure have to be `content/d1/tutorials/<folder>/index.md`?

I just checked and yes, ideally it should be in a folder! @nikolasburk

nikolasburk commented:
Thanks for the feedback @vy-ton and @Maddy-Cloudflare, I moved the tutorial into its own directory 🙏

Maddy-Cloudflare commented:
> Thanks for the feedback @vy-ton and @Maddy-Cloudflare, I moved the tutorial into its own directory 🙏

This link returns  a `404`, do you have an updated link so I can fix it?
